cpp/src/umap/simpl_set_embed/algo.h

```diff
@@ -131,23 +130,163 @@ __device__ __host__ double attractive_grad(double dist_squared,
  * randomly.
  */
 template <typename T, int TPB_X>
-__global__ void optimize_batch_kernel(
+__global__ void optimize_batch_kernel_on_shared_mem(
   T *head_embedding, int head_n, T *tail_embedding, int tail_n, const int *head,
   const int *tail, int nnz, T *epochs_per_sample, int n_vertices,
   bool move_other, T *epochs_per_negative_sample,
   T *epoch_of_next_negative_sample, T *epoch_of_next_sample, double alpha,
   int epoch, double gamma, uint64_t seed, UMAPParams params) {
+  extern __shared__ T embedding_updates[];
+
+  int row = (blockIdx.x * TPB_X) + threadIdx.x;
+  if (row < nnz) {
+    /**
+     * Positive sample stage (attractive forces)
+     */
+
+    if (epoch_of_next_sample[row] <= epoch) {
+      int j = head[row];
+      int k = tail[row];
+
+      T *current = head_embedding + (j * params.n_components);
+      T *other = tail_embedding + (k * params.n_components);
+      T *current_buffer, *other_buffer;
+
+      if (move_other) {
+        current_buffer =
+          embedding_updates + ((threadIdx.x * 2) * params.n_components);
+        other_buffer =
+          embedding_updates + (((threadIdx.x * 2) + 1) * params.n_components);
+      } else {
+        current_buffer =
+          embedding_updates + (threadIdx.x * params.n_components);
+      }
+
+      // initialization of shared memory
+      for (int d = 0; d < params.n_components; d++) {
+        current_buffer[d] = 0;
+      }
+      if (move_other) {
+        for (int d = 0; d < params.n_components; d++) {
+          other_buffer[d] = 0;
+        }
+      }
+      __syncthreads();
+
+      double dist_squared = rdist(current, other, params.n_components);
+
+      // Attractive force between the two vertices, since they
+      // are connected by an edge in the 1-skeleton.
+      double attractive_grad_coeff = 0.0;
+      if (dist_squared > 0.0) {
+        attractive_grad_coeff = attractive_grad(dist_squared, params);
+      }
+
+      /**
+       * Apply attractive force between `current` and `other`
+       * by updating their 'weights' to please them relative
+       * to their weight in the 1-skeleton.
+       * (update `other` embedding only if we are
+       * performing unsupervised training).
+       */
+      for (int d = 0; d < params.n_components; d++) {
+        double grad_d =
+          clip(attractive_grad_coeff * (current[d] - other[d]), -4.0f, 4.0f);
+
+        grad_d *= alpha;
+
+        atomicAdd(current_buffer + d, grad_d);
+
+        // happens only during unsupervised training
+        if (move_other) {
+          atomicAdd(other_buffer + d, -grad_d);
+        }
```

Comment:
`current_buffer` and `other_buffer` seem to be per-thread arrays. If so, this can just be replaced with a summation operation, no?

Reply:
Yes, absolutely. I don't know why but I was still considering them to be concurrently written on. Made me think that the first __syncthreads() is actually useless. Thanks!